### PR TITLE
Update UFL element interface: move elements to FInAT

### PR DESCRIPTION
**Pull request by mscroggs**
This is an extension of https://github.com/firedrakeproject/firedrake/pull/3122.

In https://github.com/FInAT/FInAT/pull/111, I moved the elements from `ufl.legacy` to `finat.ufl`. This PR updates tsfc to use the elements from there.

This is all a consequence of https://github.com/FEniCS/ufl/pull/197.

(I've left https://github.com/firedrakeproject/firedrake/pull/3122 up too for now in case this extension is not wanted.)

This PR is coupled with https://github.com/firedrakeproject/ufl/pull/46, ~https://github.com/firedrakeproject/tsfc/pull/302 (merged)~, ~https://github.com/FInAT/FInAT/pull/111 (merged)~, and ~https://github.com/NGSolve/ngsPETSc/pull/7 (merged)~.

### CODE REVIEW

File: requirements-git.txt
Comment (dham) [inline requirements-git.txt:1]: this is wrong

File: tests/extrusion/test_2d_cohomology.py
Comment (dham) [inline tests/extrusion/test_2d_cohomology.py:116]: This is a breaking public API change. We can't do this to users without warning.

File: tests/output/test_io_mesh.py
Comment (dham) [inline tests/output/test_io_mesh.py:106]: Looks like debugging code.

**Review by dham — CHANGES_REQUESTED** [review]
There are minor things but there is also `HDiv -> HDivElement` which is a breaking API change that we can't make without warning.

File: tests/extrusion/test_2d_cohomology.py
Comment (mscroggs) [reply tests/extrusion/test_2d_cohomology.py]: I think I've sucessfully reverted this change. I was getting issues somewhere due to `ufl.HDiv` being the Sobolev space, but I'm not sure why I changed all these rather than making sure the function `HDiv` was defined/imported later

Comment (mscroggs) [reply tests/extrusion/test_2d_cohomology.py]: Nevermind, I see the issue: You used to be able to call `ufl.HDiv` but that got removed in https://github.com/FEniCS/ufl/pull/197 when we deprecated `HDivElement`. Update to finat incoming...

Comment (mscroggs) [reply tests/extrusion/test_2d_cohomology.py]: I think https://github.com/FInAT/FInAT/pull/111/commits/666e924899c063254655e8b54bf937a63a715b6e should restore the old behaviour of HDiv and HCurl

Comment (mscroggs) [reply tests/extrusion/test_2d_cohomology.py]: I've opened an issue in case you want to remove my hack to keep this syntax at some point: https://github.com/firedrakeproject/firedrake/issues/3213

File: tests/slate/test_mimetic.py
Comment (mscroggs) [inline tests/slate/test_mimetic.py:46]:
```suggestion
    product_elt = HCurl(TensorProductElement(horiz_elt, vert_elt))
```

File: tests/regression/test_fs_caching.py
Comment (mscroggs) [inline tests/regression/test_fs_caching.py:139]:
```suggestion
    W2 = FunctionSpace(e, HCurl(W0))
```

File: tests/regression/test_fs_caching.py
Comment (mscroggs) [inline tests/regression/test_fs_caching.py:223]:
```suggestion
    V2 = FunctionSpace(e, HCurl(V0))
```

File: tests/regression/test_fs_caching.py
Comment (mscroggs) [inline tests/regression/test_fs_caching.py:251]:
```suggestion
    V2 = FunctionSpace(e, HCurl(V0))
```

**Comment by mscroggs** [comment]
I've fixed the behaviour of `HDiv` so that the change to `HDivElement` is no longer necessary. Hopefully this is ready to go now

File: requirements-git.txt
Comment (ReubenHill) [inline requirements-git.txt:4]: Branch selection is meant to be done in `build.yml` as is done [here](https://github.com/firedrakeproject/firedrake/pull/3193/files). This could be why doc build is failing (@JDBetteridge  you're the expert there!) We can't merge whilst we have failing doc build. It would be handy to change this and retrigger tests before our meeting at 1600 UTC today where we will hopefully get this all merged

File: scripts/firedrake-install
Comment (connorjward) [inline scripts/firedrake-install:1430]: This should be looked at.

File: tests/extrusion/test_real_tensorproduct.py
Comment (connorjward) [inline tests/extrusion/test_real_tensorproduct.py:76]: These tolerances are already tiny so I guess this isn't an issue. I am surprised that this is required for this PR to land though.

File: tests/output/test_io_mesh.py
Comment (connorjward) [inline tests/output/test_io_mesh.py:150]: This looks wrong.

**Review by connorjward — COMMENTED** [review]
99% fine. Happy to approve once my comments are addressed.

File: tests/output/test_io_mesh.py
Comment (mscroggs) [reply tests/output/test_io_mesh.py:150]: @ksagiyam suggested over in https://github.com/firedrakeproject/firedrake/pull/3122 that this test was never meant to pass in parallel

Comment (connorjward) [reply tests/output/test_io_mesh.py:150]: Ah OK!

File: scripts/firedrake-install
Comment (mscroggs) [reply scripts/firedrake-install:1430]: This is the branch that https://github.com/NGSolve/ngsPETSc/pull/7 is suggesting to merge. Should be reset once that merge is done

File: tests/extrusion/test_real_tensorproduct.py
Comment (mscroggs) [reply tests/extrusion/test_real_tensorproduct.py]: I've changed this back. Let's see if the CI run passes now...

File: requirements-git.txt
Comment (mscroggs) [reply requirements-git.txt]: I corrected how branch selection is done, docs still fail. Further suggestions welcome, I have no idea what I'm doing with the docs build

Comment (connorjward) [reply requirements-git.txt]: This may not actually be a problem. We build the docs in containers so if we modify dependencies these changes are not known about by the container and then things won't resolve until we rebuild the containers.

Comment (JDBetteridge) [reply requirements-git.txt]: Don't worry Matt, I would expect our docs build to fail. It is currently set to build in a container with everything pre-installed at the default branch. We currently do not have a good (automatic) way to change the branch of dependencies when the docs build. If you can confirm that docs build locally (`make linkcheck && make html && make latexpdf` in the docs directory) that's more than what we insist on for most PRs!

Comment (ReubenHill) [reply requirements-git.txt]: You'd need to run `firedrake-install --documentation-dependencies` first and move to the `docs` directory before running `make`. Also I've never managed to get `make latexpdf` working on my mac so I reckon as long as `make linkcheck && make html` works you're good to go.

Comment (JDBetteridge) [reply requirements-git.txt]: TBF we don't _need_ the LaTeX docs to build on mac, so long as they build on CI we can release PDF manuals.

File: tests/extrusion/test_real_tensorproduct.py
Comment (mscroggs) [reply tests/extrusion/test_real_tensorproduct.py]: Seems this change was necessary: https://github.com/firedrakeproject/firedrake/actions/runs/6876364089/job/18701826166?pr=3166

File: scripts/firedrake-install
Comment (mscroggs) [reply scripts/firedrake-install:1430]: I can't see any way to get firedrake-install to use a branch of ngsPETSc without this change

File: requirements-git.txt
Comment (mscroggs) [reply requirements-git.txt]: `make html` and `makelatexpdf` succeed locally.
`make linkcheck` complains that https://www.maths.ox.ac.uk/people/fabian.laakmann gives a 403 error. I guess this is unrelated to these changes

Comment (JDBetteridge) [reply requirements-git.txt]: Thanks, that link is working intermittently, so if it's a problem on CI later we'll fix it in another PR

File: scripts/firedrake-install
Comment (JDBetteridge) [reply scripts/firedrake-install:1430]: This is sensible, I will poke @UZerbinati who might be able to give us a status update.

File: tests/extrusion/test_real_tensorproduct.py
Comment (connorjward) [reply tests/extrusion/test_real_tensorproduct.py:76]: OK. I think this is fine then.

File: scripts/firedrake-install
Comment (mscroggs) [reply scripts/firedrake-install]: https://github.com/NGSolve/ngsPETSc/pull/7 is merged, so resolving this

File: .github/workflows/build.yml
Comment (JDBetteridge) [inline .github/workflows/build.yml:84]:
```suggestion
```

File: .github/workflows/build.yml
Comment (JDBetteridge) [inline .github/workflows/build.yml:93]:
```suggestion
```

**Review by JDBetteridge — CHANGES_REQUESTED** [review]
Don't worry about adding these I'll add the suggestions to a batch once I've checked the other branches :slightly_smiling_face:

**Review by JDBetteridge — APPROVED** [review]

**Comment by JDBetteridge** [comment]
Thanks for all your hard work on this @mscroggs !

**Pull request merged**
